File: data/tomoe-0.6.0/test/test_char.rb
require 'tomoe-test-utils'

class TestChar < Test::Unit::TestCase
  include TomoeTestUtils

  def test_dump_without_data
    char = Tomoe::Char.new
    assert_equal("", char.to_xml)
  end

  def test_dump_with_utf8
    char = Tomoe::Char.new
    char.utf8 = "あ"
    xml = "  <character>\n"
    xml << "    <utf8>あ</utf8>\n"
    xml << "  </character>\n"
    assert_equal(xml, char.to_xml)
  end

  def test_dump_with_variant
    hashigo_daka = ucs4_to_utf8(39641) # はしご高
    char = Tomoe::Char.new
    char.utf8 = "高"
    char.variant = hashigo_daka
    xml = "  <character>\n"
    xml << "    <utf8>高</utf8>\n"
    xml << "    <variant>#{hashigo_daka}</variant>\n"
    xml << "  </character>\n"
    assert_equal(xml, char.to_xml)
  end

  def test_dump_with_radicals
    sanzui = ucs4_to_utf8(27701) # さんずい
    char = Tomoe::Char.new
    char.utf8 = "池"
    char.add_radical(sanzui)
    xml = "  <character>\n"
    xml << "    <utf8>池</utf8>\n"
    xml << "    <radicals>\n"
    xml << "      <radical>#{sanzui}</radical>\n"
    xml << "    </radicals>\n"
    xml << "  </character>\n"
    assert_equal(xml, char.to_xml)
  end

  def test_load_from_dumped_xml
    char = Tomoe::Char.new
    char.utf8 = "あ"
    new_char = Tomoe::Char.new(char.to_xml)
    assert_equal(char.utf8, new_char.utf8)
  end

  def test_load_from_dumped_xml_with_variant
    hashigo_daka = ucs4_to_utf8(39641) # はしご高
    char = Tomoe::Char.new
    char.utf8 = "高"
    char.variant = hashigo_daka

    new_char = Tomoe::Char.new(char.to_xml)
    assert_equal(char.utf8, new_char.utf8)
    assert_equal(hashigo_daka, new_char.variant)
  end

  def test_load_from_dumped_xml_with_radicals
    sanzui = ucs4_to_utf8(27701) # さんずい
    char = Tomoe::Char.new
    char.utf8 = "池"
    char.add_radical(sanzui)

    new_char = Tomoe::Char.new(char.to_xml)
    assert_equal(char.utf8, new_char.utf8)
    assert_equal([sanzui], new_char.radicals)
  end

  def test_load_from_dumped_xml_with_n_strokes
    utf8 = "a"
    n_strokes = 5

    char = Tomoe::Char.new
    char.utf8 = utf8
    assert_equal(-1, char.n_strokes)
    char.n_strokes = n_strokes
    assert_equal(n_strokes, char.n_strokes)

    new_char = Tomoe::Char.new(char.to_xml)
    assert_equal(utf8, new_char.utf8)
    assert_equal(n_strokes, new_char.n_strokes)
  end

  def test_n_strokes_accessor
    char = Tomoe::Char.new
    assert_equal(-1, char.n_strokes)

    char.n_strokes = 5
    assert_equal(5, char.n_strokes)
  end

  def test_variant_accessor
    variant = "異"

    char = Tomoe::Char.new
    assert_nil(char.variant)

    char.variant = variant
    assert_equal(variant, char.variant)
  end

  def test_writing_accessor
    strokes = [[[0, 0], [10, 10]]]
    writing = Tomoe::Writing.new
    writing.move_to(*strokes[0][0])
    writing.line_to(*strokes[0][1])
    assert_equal(strokes, writing.strokes)

    char = Tomoe::Char.new
    assert_nil(char.writing)

    char.writing = writing
    assert_equal(strokes, char.writing.strokes)
  end

  def test_reading_accessor
    reading_a = Tomoe::Reading.new(Tomoe::Reading::JA_ON, "あ")

    reading_kai = Tomoe::Reading.new(Tomoe::Reading::JA_KUN, "カイ")

    char = Tomoe::Char.new
    assert_equal([], char.readings)

    char.add_reading(reading_a)
    assert_equal([reading_a.reading].sort,
                 char.readings.collect do |reading|
                   reading.reading
                 end.sort)

    char.add_reading(reading_kai)
    assert_equal([reading_a.reading, reading_kai.reading].sort,
                 char.readings.collect do |reading|
                   reading.reading
                 end.sort)
  end

  def test_radical_accessor
    radical1 = "木"
    radical2 = "水"

    char = Tomoe::Char.new
    assert_equal([], char.radicals)

    char.add_radical(radical1)
    assert_equal([radical1], char.radicals.sort)

    char.add_radical(radical2)
    assert_equal([radical1, radical2].sort, char.radicals.sort)
  end

  def test_comparable
    char1 = Tomoe::Char.new
    char1.utf8 = "あ"
    char2 = Tomoe::Char.new
    char2.utf8 = "あ"
    assert_equal(char1, char2)
  end
end

File: data/tomoe-0.6.0/test/python/test_char.py
# encoding: utf-8
import os
import sys
import unittest
import tomoe

class TomoeCharTest(unittest.TestCase):

    def testDuplicate(self):
        data = """
<character>
  <utf8>池</utf8>
  <number-of-strokes>6</number-of-strokes>
  <readings>
    <reading type="ja_on">タ</reading>
    <reading type="ja_on">チ</reading>
    <reading type="ja_kun">いけ</reading>
    <reading type="ja_unknown">あんのうん</reading>
  </readings>
  <meta>
    <jis208>35-51</jis208>
    <ucs>6c60</ucs>
    <jouyou>2</jouyou>
    <meaning>pond, cistern, pool, reservoir</meaning>
  </meta>
</character>
        """
        char1 = tomoe.tomoe_char_new_from_xml_data(data, int(-1))
        self.assertNotEqual(None, char1)
        char2 = char1.dup()
        self.assertEqual(char1.to_xml(), char2.to_xml())

    def testUTF8(self):
        utf8 = '池'

        tomoe_char = tomoe.Char()
        self.assertEqual(None, tomoe_char.get_utf8())
        tomoe_char.set_utf8(utf8)
        self.assertEqual(utf8, tomoe_char.get_utf8())

        tomoe_char = tomoe.Char(utf8 = utf8)
        self.assertEqual(utf8, tomoe_char.get_utf8())

        new_utf8 = '地'
        tomoe_char.set_utf8(new_utf8)
        self.assertEqual(new_utf8, tomoe_char.get_utf8())

    def testStrokes(self):
        n_strokes = 2
        tomoe_char = tomoe.Char()
        self.assertEqual(-1, tomoe_char.get_n_strokes())
        tomoe_char.set_n_strokes(n_strokes)
        self.assertEqual(n_strokes, tomoe_char.get_n_strokes())

        tomoe_char = tomoe.Char(n_strokes = n_strokes)
        self.assertEqual(n_strokes, tomoe_char.get_n_strokes())

        n_strokes = 3
        tomoe_char.set_n_strokes(n_strokes)
        self.assertEqual(n_strokes, tomoe_char.get_n_strokes())

    def testReading(self):
        reading_string = 'いけ'
        reading1 = tomoe.Reading(tomoe.READING_UNKNOWN, reading_string)

        tomoe_char = tomoe.Char()
        self.assertEqual(0, len(tomoe_char.get_readings()))

        tomoe_char.add_reading(reading1)
        self.assertEqual([reading1], tomoe_char.get_readings())

        reading2 = tomoe.Reading(tomoe.READING_JA_KUN, reading_string)
        tomoe_char.add_reading(reading2)
        self.assertEqual([reading1, reading2].sort(), tomoe_char.get_readings().sort())

    def testRadical(self):
        radical_string1 = 'いけ'

        tomoe_char = tomoe.Char()
        self.assertEqual(0, len(tomoe_char.get_radicals()))

        tomoe_char.add_radical(radical_string1)
        self.assertEqual([radical_string1], tomoe_char.get_radicals())

        radical_string2 = 'ほげ'
        tomoe_char.add_radical(radical_string2)
        self.assertEqual([radical_string1, radical_string2].sort(), tomoe_char.get_radicals().sort())

    def testWriting(self):
        writing = tomoe.Writing()

        tomoe_char = tomoe.Char()
        self.assertEqual(None, tomoe_char.get_writing())
        tomoe_char.set_writing(writing)
        self.assertEqual(writing, tomoe_char.get_writing())

        tomoe_char = tomoe.Char(writing = writing)
        self.assertEqual(writing, tomoe_char.get_writing())

        writing.move_to(1, 1)
        writing.line_to(2, 2)
        tomoe_char.set_writing(writing)
        self.assertEqual(writing.to_xml(), tomoe_char.get_writing().to_xml())

    def testVariant(self):
        variant = '池'
        tomoe_char = tomoe.Char()
        self.assertEqual(None, tomoe_char.get_variant())
        tomoe_char.set_variant(variant)
        self.assertEqual(variant, tomoe_char.get_variant())

        tomoe_char = tomoe.Char(variant = variant)
        self.assertEqual(variant, tomoe_char.get_variant())

        variant = '地'
        tomoe_char.set_variant(variant)
        self.assertEqual(variant, tomoe_char.get_variant())

    def testCompare(self):
        char1 = '池'
        char2 = '地'
        
        tomoe_char1 = tomoe.Char(utf8 = char1)
        tomoe_char2 = tomoe.Char(utf8 = char1)
        self.assertEqual(tomoe_char1, tomoe_char2)
        
        tomoe_char2 = tomoe.Char(utf8 = char2)
        self.assertNotEqual(0, tomoe_char1, tomoe_char2)

# vi:ts=4:nowrap:ai:expandtab
